prefect_utils/snowflake.py

```diff
@@ -194,3 +194,140 @@ def load_ga_data_to_snowflake(
+        SELECT 1 FROM {table}
+        WHERE date(PROPERTIES:{date_property})='{date}'
+        """.format(
+            table=qualified_table_name(sf_database, sf_schema, sf_table),
+            date=date,
+            date_property=date_property,
+        )
+        cursor = sf_connection.cursor()
+        cursor.execute(query)
+        row = cursor.fetchone()
+    except snowflake.connector.ProgrammingError as e:
+        if "does not exist" in e.msg:
+            # If so then the query failed because the table doesn't exist.
+            row = None
+        else:
+            raise
+
+    if row and not overwrite:
+        return
+
+    try:
+        # Create the generic loading table
+        query = """
+        CREATE TABLE IF NOT EXISTS {table} (
+            ID NUMBER AUTOINCREMENT START 1 INCREMENT 1,
+            LOAD_TIME TIMESTAMP_LTZ DEFAULT CURRENT_TIMESTAMP(),
+            ORIGIN_FILE_NAME VARCHAR(16777216),
+            ORIGIN_FILE_LINE NUMBER(38,0),
+            ORIGIN_STR VARCHAR(16777216),
+            PROPERTIES VARIANT
+        );
+        """.format(
+            table=qualified_table_name(sf_database, sf_schema, sf_table)
+        )
+
+        sf_connection.cursor().execute(query)
+
+        # Delete existing data in case of overwrite.
+        if overwrite:
+            query = """
+            DELETE FROM {table}
+            WHERE date(PROPERTIES:{date_property})='{date}'
+            """.format(
+                table=qualified_table_name(sf_database, sf_schema, sf_table),
+                date=date,
+                date_property=date_property,
+            )
+            sf_connection.cursor().execute(query)
+
+        stage_url = 's3://{bucket}/{path}'.format(bucket=s3_bucket, path=s3_key_path)
+
+        # Create stage
+        query = """
+        CREATE OR REPLACE STAGE {stage_name}
+            URL = '{stage_url}'
+            STORAGE_INTEGRATION = {storage_integration}
+            FILE_FORMAT = (TYPE = JSON STRIP_OUTER_ARRAY = TRUE);
+        """.format(
+            stage_name=qualified_stage_name(sf_database, sf_schema, sf_table),
+            stage_url=stage_url,
+            storage_integration=sf_storage_integration_name,
+        )
+        sf_connection.cursor().execute(query)
+
+        query = """
+        COPY INTO {table} (origin_file_name, origin_file_line, origin_str, properties)
+            FROM (
+                SELECT
+                    metadata$filename,
+                    metadata$file_row_number,
+                    t.$1,
+                    CASE
+                        WHEN CHECK_JSON(t.$1) IS NULL THEN t.$1
+                        ELSE NULL
+                    END
+                FROM @{stage_name} t
+            )
+        PATTERN='{pattern}'
+        FORCE={force}
+        """.format(
+            table=qualified_table_name(sf_database, sf_schema, sf_table),
+            stage_name=qualified_stage_name(sf_database, sf_schema, sf_table),
+            pattern=pattern,
```

Comment:
So right now this looks like it will ingest all data from this path that match the pattern, not just the file for the given date. That seems likely to become a race condition to me, given that everything else happening here is date-dependent, especially as this is mapped and we may have many of these running simultaneously with the date deletes. Does it make sense to extend `pattern` to add the date here?

Reply:
The stage url is currently set to
`stage_url = 's3://{bucket}/{path}'.format(bucket=s3_bucket, path=s3_key_path)`
So the COPY command will only load data from that path, whether that's a folder or a file. In case of a folder, everything under it, in case of a file path, only the single file. 

For affiliate window and cybersource reports, `s3_key_path` is the complete path of the file(minus the bucket). 
So this will only load the single file.

Reply:
I think that also won't work as the various jobs will be using the same stage name, force recreating it with different urls, right? I'm not sure how transactions work with stages.

Reply:
Stage name is down to the table level. Unless there are multiple jobs writing to the same table, should not be a problem. We could append a timestamp or something to the stage name, but we should also clean up after the loading is done if we choose to do that. 
```
def qualified_stage_name(database, schema, table) -> str:
    """
    Fully qualified Snowflake stage name.
    """
    return "{database}.{schema}.{table}_stage".format(
        database=database, schema=schema, table=table,
    )

Reply:
If we're running multiple days they will all map and run this at the same time, right?

Reply:
Good point, although currently we don't have anything set up to support tasks running in parallel. Everything runs sequentially right now.

Reply:
I think we need to setup dask on kubernetes to support parallel running tasks.

Reply:
I imagine we'll need that on our roadmap, so it's worth planning for if multiple jobs will be using this. Wouldn't it make sense to point the stage to the directory and the COPY INTO at the specific file? Then we could also make the stage `CREATE IF NOT EXISTS...`? Or do we still have some reason to always recreate it?

Reply:
Just wondering if there is much value in persisting the stage, if it's no longer needed for keeping state about what was loaded.  So rather than working to create a single stage that is shared across multiple tables, should we try to have as few stages persist as possible, and just create them as needed?

Reply:
I'd ask the opposite question -- is there a reason to not always create the stage?  If we don't need to persist the stages defined on each table, we could use alternative naming for them, and then delete the stage once it's done.   So appending a timestamp is one way to do that.   There are also other ways of naming stages.  (Not remembering the exact details, but I used one of them in my stuff for loading data to Snowflake from notebook dataframes.)

Reply:
This is true, we should be able to copy into without a stage just using the storage integration. https://docs.snowflake.com/en/sql-reference/sql/copy-into-table.html#loading-files-directly-from-an-external-location

Reply:
Thanks for looking that up. I think I like this better, that is remove the stage creation step entirely.

Reply:
Okay, I was just looking at the docs for COPY into. It looks like external locations are not usable with transformation, like we do here.

Reply:
```
COPY INTO [<namespace>.]<table_name>
     FROM { internalStage | externalStage | externalLocation }
[ FILES = ( '<file_name>' [ , '<file_name>' ] [ , ... ] ) ]
[ PATTERN = '<regex_pattern>' ]
[ FILE_FORMAT = ( { FORMAT_NAME = '[<namespace>.]<file_format_name>' |
                    TYPE = { CSV | JSON | AVRO | ORC | PARQUET | XML } [ formatTypeOptions ] } ) ]
[ copyOptions ]
[ VALIDATION_MODE = RETURN_<n>_ROWS | RETURN_ERRORS | RETURN_ALL_ERRORS ]

/* Data load with transformation */
COPY INTO [<namespace>.]<table_name> [ ( <col_name> [ , <col_name> ... ] ) ]
     FROM ( SELECT [<alias>.]$<file_col_num>[.<element>] [ , [<alias>.]$<file_col_num>[.<element>] ... ]
            FROM { internalStage | externalStage } )
[ FILES = ( '<file_name>' [ , '<file_name>' ] [ , ... ] ) ]
[ PATTERN = '<regex_pattern>' ]
[ FILE_FORMAT = ( { FORMAT_NAME = '[<namespace>.]<file_format_name>' |
                    TYPE = { CSV | JSON | AVRO | ORC | PARQUET | XML } [ formatTypeOptions ] } ) ]
[ copyOptions ]
[ VALIDATION_MODE = RETURN_<n>_ROWS | RETURN_ERRORS | RETURN_ALL_ERRORS ]
```

Reply:
I changed the stage name by appending a uuid. Thoughts ?
```
stage_name_with_uuid = qualified_stage_name(sf_database, sf_schema, sf_table) +
    '_{}'.format(uuid.uuid4().hex)

Reply:
I think this would result in huge amounts of orphaned stages. I don't think we need to do anything with the stage name if we:

- CREATE IF NOT EXISTS on the stage
- Add the name of the file we're loading to the COPY INTO pattern

I know we don't have the filename pattern in here yet, but I think it makes sense to be explicit about what we're loading if we're only doing one date / file at at time.

Reply:
Sorry, I meant to drop the stage after COPY INTO.

Reply:
Regarding the pattern, the task has a `pattern` parameter(used in COPY INTO statement) that can be passed down when the task is called in the workflow. 

The current implementation defines the stage url as 
`stage_url = 's3://{bucket}/{path}'.format(bucket=s3_bucket, path=s3_key_path)`
where 
`s3_bucket (str): S3 bucket from which to load data from.`
`s3_key_path (str): Path of the S3 objects relative to s3_bucket.`
and with the optional 
`pattern (str, optional): Path pattern/regex to match S3 objects to copy. Defaults to .*.`
which is specified in the COPY command. 

`s3_key_path` can be a path to a directory or a file and then we also have the `pattern` parameter. 

Doesn't that give us enough control if we want to be explicit about what we're loading ?

Reply:
Some more thoughts on this
> 
> * CREATE IF NOT EXISTS on the stage
> * Add the name of the file we're loading to the COPY INTO pattern
> 
> I know we don't have the filename pattern in here yet, but I think it makes sense to be explicit about what we're loading if we're only doing one date / file at at time.

You're suggesting that we have a single stage for that table, concurrent same tasks would then share the stage and COPY would be controlled by either a list of files passed to the COPY command or through a pattern. 

In order to do that, we would be creating the stage on the parent directory once(CREATE IF NOT EXISTS), i.e `s3://bucket/prod/warehouse/fees/affiliate_window_prefect/`

And then we would also pass in FILES or use the pattern to COPY to control what gets copied into snowflake, a single file or date pattern for these kind of reports. In case of pattern, I'm assuming snowflake lists all objects under the stage url and then does pattern matching, for FILES I think it just appends it to the stage url.

Reply:
Yes, that's my suggestion. It just makes for less messing with the stage, which could impact other processes. If we can just pass in the file to `s3_key_path` I'm all for it, but we should make sure the param is clear that we expect a file there.

Reply:
For usage reference.
```
    # Step 2: Upload each report to S3
    uploaded_files = write_report_to_s3.map(
        report_info=fetched_reports,
        s3_bucket=unmapped(s3_bucket),
        s3_path=unmapped(s3_path),
    )
Here `uploaded_files` is a list of s3 prefixes relative to the bucket, 
`prod/warehouse/fees/affiliate_window_prefect/2020-01-01/2020-01-01.json`

It is easier to just pass this list to the snowflake loading task without having to play around with it.